diff --git a/CMakeLists.txt b/CMakeLists.txt
index d13fba9d8..155d9d160 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -48,7 +48,7 @@ set_package_properties(KF6KirigamiAddons PROPERTIES
 )
 
 pkg_check_modules(Flatpak IMPORTED_TARGET flatpak>=0.11.8)
-pkg_check_modules(Fwupd IMPORTED_TARGET fwupd>=2.1.1)
+pkg_check_modules(Fwupd IMPORTED_TARGET fwupd>=1.9.4)
 pkg_check_modules(Markdown IMPORTED_TARGET libmarkdown)
 if(Markdown_FOUND AND Markdown_VERSION VERSION_GREATER_EQUAL 3)
     add_definitions(-DMARKDOWN3)
diff --git a/libdiscover/backends/FwupdBackend/FwupdBackend.cpp b/libdiscover/backends/FwupdBackend/FwupdBackend.cpp
index f0e35d5b4..8237157ff 100644
--- a/libdiscover/backends/FwupdBackend/FwupdBackend.cpp
+++ b/libdiscover/backends/FwupdBackend/FwupdBackend.cpp
@@ -245,8 +245,11 @@ FwupdResource *FwupdBackend::createApp(FwupdDevice *device)
 void FwupdBackend::handleError(GError *perror)
 {
     // TODO: localise the error message
-    if (perror && !g_error_matches(perror, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE) && !g_error_matches(perror, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED)
-        && !g_error_matches(perror, FWUPD_ERROR, FWUPD_ERROR_NOTHING_TO_DO)) {
+    if (perror && !g_error_matches(perror, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE)
+        && !g_error_matches(perror, FWUPD_ERROR, FWUPD_ERROR_NOTHING_TO_DO)
+        // "Systemd service is masked" error, which is not really an error, but has no clean GError string,
+        // so we have to compare codes and strings. Thankfully it's unlocalized!
+        && !(perror->code == 3 && QString::fromUtf8(perror->message).contains(QStringLiteral("unit is masked")))) {
         const QString msg = QString::fromUtf8(perror->message);
         QTimer::singleShot(0, this, [this, msg]() {
             Q_EMIT passiveMessage(msg);
